File: backend/app/db/models/packaging.py
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.db.models.logistics import ShipmentPackage
    from app.db.models.procurement import PurchaseAllocation


class Package(Base):
    __tablename__ = "packages"

    package_id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    package_no: Mapped[str] = mapped_column(String, unique=True, nullable=False)
    status: Mapped[str] = mapped_column(String, nullable=False)

    items: Mapped[list["PackageItem"]] = relationship(back_populates="package")
    shipment_links: Mapped[list["ShipmentPackage"]] = relationship(back_populates="package")


class PackageItem(Base):
    """A packed quantity of stock set aside for one customer order (a
    purchase_allocation), not a raw product+purchase. A package can therefore
    hold items sourced from several purchases and destined for several
    customer orders — packing follows the shipment/destination, not any
    single purchase order."""

    __tablename__ = "package_items"

    package_item_id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    package_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("packages.package_id"), nullable=False
    )
    purchase_allocation_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("purchase_allocations.allocation_id"), nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)

    package: Mapped["Package"] = relationship(back_populates="items")
    allocation: Mapped["PurchaseAllocation"] = relationship(back_populates="package_items")

File: frontend/src/data/fixtures/packaging.ts
import type { Package, PackageItem } from "@/data/types";

export const packages: Package[] = [
  { packageId: "PK-000", packageNo: "PK-000", status: "shipped" },
  { packageId: "PK-001", packageNo: "PK-001", status: "shipped" },
  { packageId: "PK-002", packageNo: "PK-002", status: "preparing" },
  { packageId: "PK-003", packageNo: "PK-003", status: "shipped" },
  { packageId: "PK-004", packageNo: "PK-004", status: "preparing" },
  { packageId: "PK-005", packageNo: "PK-005", status: "shipped" },
  { packageId: "PK-006", packageNo: "PK-006", status: "shipped" },
];

// Each line references the specific purchase-to-order allocation it was
// packed for (see procurement.ts) — not a raw product. That's what lets a
// single package span multiple purchases and multiple customer orders:
// PK-002 packs stock from both PO-002 (for SO-1025) and PO-007 (for
// SO-1050); PK-004 packs the same purchased batch of Khaki Pants split
// across two different customers, SO-1035 and SO-1050.
export const packageItems: PackageItem[] = [
  { packageItemId: "PKI-000-1", packageId: "PK-000", purchaseAllocationId: "PA-1", quantity: 60 },
  { packageItemId: "PKI-000-2", packageId: "PK-000", purchaseAllocationId: "PA-2", quantity: 20 },

  { packageItemId: "PKI-001-1", packageId: "PK-001", purchaseAllocationId: "PA-3", quantity: 40 },
  { packageItemId: "PKI-001-2", packageId: "PK-001", purchaseAllocationId: "PA-4", quantity: 30 },

  { packageItemId: "PKI-002-1", packageId: "PK-002", purchaseAllocationId: "PA-5", quantity: 60 },
  { packageItemId: "PKI-002-2", packageId: "PK-002", purchaseAllocationId: "PA-6", quantity: 40 },
  { packageItemId: "PKI-002-3", packageId: "PK-002", purchaseAllocationId: "PA-17", quantity: 20 },

  { packageItemId: "PKI-003-1", packageId: "PK-003", purchaseAllocationId: "PA-7", quantity: 30 },
  { packageItemId: "PKI-003-2", packageId: "PK-003", purchaseAllocationId: "PA-8", quantity: 15 },
  { packageItemId: "PKI-003-3", packageId: "PK-003", purchaseAllocationId: "PA-9", quantity: 10 },

  { packageItemId: "PKI-004-1", packageId: "PK-004", purchaseAllocationId: "PA-10", quantity: 50 },
  { packageItemId: "PKI-004-2", packageId: "PK-004", purchaseAllocationId: "PA-16", quantity: 25 },
  { packageItemId: "PKI-004-3", packageId: "PK-004", purchaseAllocationId: "PA-11", quantity: 100 },

  { packageItemId: "PKI-005-1", packageId: "PK-005", purchaseAllocationId: "PA-12", quantity: 70 },
  { packageItemId: "PKI-005-2", packageId: "PK-005", purchaseAllocationId: "PA-13", quantity: 35 },

  { packageItemId: "PKI-006-1", packageId: "PK-006", purchaseAllocationId: "PA-14", quantity: 60 },
  { packageItemId: "PKI-006-2", packageId: "PK-006", purchaseAllocationId: "PA-15", quantity: 45 },
];
